gamification/models.py

```diff
@@ -0,0 +1,182 @@
+                                this activity was performed.
+        """
+        self.update_score_and_level(points)
+        self.add_activity(points,
+                          activity_string,
+                          performed_at,
+                          updated_at,
+                          )
+
+    def deduct_points(self, points_to_deduct, activity_string,
+                      performed_at, updated_at):
+        """
+        Deduct points for performing some specific activities.
+        """
+        self.add_points(points_to_deduct,
+                        activity_string,
+                        performed_at,
+                        updated_at,
+                        )
+
+    def find_level_for_score(self, score):
+        """
+        Find suitable level based on the total score earned.
+        """
+        level = Level.objects.get(min_score__lte=score, max_score__gt=score)
+        return level
+
+    def update_score_and_level(self, points):
+        """
+        Update score and level based on points.
+        """
+        if points < 0 and self.score < abs(points):
+            self.score = 0
+        else:
+            self.score += points
+
+        self.level = self.find_level_for_score(self.score)
+        self.save()
+
+    def add_activity(self, points, activity_string, performed_at, updated_at):
+        """
+        Add a new activity to the participant.
+        """
+        activity = Activity.objects.create(
+            name=activity_string,
+            points=points,
+            performer=self,
+            performed_at=performed_at,
+            updated_at=updated_at)
+        self.activities.add(activity)
+
+    def find_badges_for_activity(self, activities):
+        """
+        Find the badge based on the activities peformed by the participant.
+
+        :param activities: a QuerySet dict containing the 'name'
+                           as key and 'name of the activity' as value.
+        :return:           a badge object.
+        """
+        activities = [activity['name'] for activity in activities]
+        badge_objects_list = []
+        badges = Badge.objects.all()
+        for badge in badges:
+            b_activities = badge.b_activities.values('name')
+            b_activities = [b_activity['name'] for b_activity in b_activities]
+            # Skip the badge which have no activities
+            if not len(b_activities):
+                continue
+            match_activity_list = []
+            for b_activity in b_activities:
+                b_activity_words_list = b_activity.split()
+                for activity in activities:
+                    if all(x in activity for x in b_activity_words_list):
+                        match_activity_list.append(1)
+                        break
+            if len(b_activities) == len(match_activity_list):
+                # The participant have performed all the activities listed
```

Comment:
add a log entry that the participant was awarded a badge

Reply:
oh wow, this log entry is quite noisy as it is being done many times.

drop it down to debug, and add a comment in data/badges.py about the two badges without activities.